internal/ethapi/api_extra.go

```diff
@@ -92,3 +92,40 @@ func (s *BlockChainAPI) GetBadBlocks(ctx context.Context) ([]*BadBlockArgs, erro
 	}
 	return results, nil
 }
+
+// isHistoricalStateQueryAllowed returns a nil error only if:
+//   - the node is configured to accept historical proofs queries; or
+//   - the block number given is within the window of recent accepted
+//     blocks and not yet part of the historical blocks.
+//
+// Otherwise, a non-nil error is returned.
+func (s *BlockChainAPI) isHistoricalStateQueryAllowed(blockNumOrHash rpc.BlockNumberOrHash) (err error) {
+	historicalProofs, recentBlocksWindow := s.b.HistoricalConfig()
+	if historicalProofs {
+		return nil
+	}
+
+	lastBlock := s.b.LastAcceptedBlock()
+	lastNumber := lastBlock.NumberU64()
+
+	var number uint64
+	if blockNumOrHash.BlockNumber != nil {
+		number = uint64(blockNumOrHash.BlockNumber.Int64())
+	} else {
+		block, err := s.b.BlockByNumberOrHash(context.Background(), blockNumOrHash)
+		if err != nil {
+			return fmt.Errorf("getting block from hash: %s", err)
+		}
+		number = block.NumberU64()
+	}
+
+	var oldestRecentNumber uint64
+	if lastNumber > recentBlocksWindow {
+		oldestRecentNumber = lastNumber - recentBlocksWindow
+	}
+	if number >= oldestRecentNumber {
+		return nil
+	}
+	return fmt.Errorf("block number %d is before the oldest non-historical block number %d (window of %d blocks)",
```

Comment:
can we return an `bool, error`? this does look like function was successfull but actually returning the result rather than an error.

Reply:
I explained why there: https://github.com/ava-labs/coreth/pull/719#discussion_r1897926695

(sorry conversation got resolved)

Otherwise, we would need to access a bunch of things at the caller level to have a useful enough error.